### PR TITLE
163639 Create top-up document template in payment-generator service

### PR DESCRIPTION
**Pull request by AbidemiAdio**

### CODE REVIEW

File: docs/asyncapi.yaml [hunk 1]
```diff
@@ -302,7 +302,7 @@ components:
     Calculated:
       type: string
       description: The calculated date of the payment request
-      example: 16 June 2022
+      example: 16 Jun 2022
```
Comment (marctemp) [inline docs/asyncapi.yaml:305]: 👍

File: test/unit/generator/content/get-address.test.js [hunk 2]
```diff
@@ -170,14 +170,14 @@ describe('get address', () => {
...
+    expect(result.absolutePosition.x).toBe(0)
   })
 
   test('returns correct location y axis', () => {
     const address = {
       postcode: 'NE1 1AA'
     }
     const result = getAddress(businessName, address)
-    expect(result.absolutePosition.y).toBe(99.225)
+    expect(result.absolutePosition.y).toBe(70.875)
```
Comment (marctemp) [inline test/unit/generator/content/get-address.test.js:181]: why isn't this the defined 25 in getAddress?

File: app/generator/content/schedule/index.js [hunk 3]
```diff
@@ -1,10 +1,12 @@
 const part1 = require('./part1')
-const summary = require('./summary')
+const part2 = require('./part2')
+const part3 = require('./part3')
 
 const createContent = (schedule) => {
```
Comment (marctemp) [inline app/generator/content/schedule/index.js:5]: part1, part2 and part3 make sense for statements as they are lit. defined that way, i.e. `"Part 1: Payment summary"`

and I get the idea of splitting documents up into sep. parts, sorta hassa polymorphic feel but it's not as clear-cut where these splits are for a schedule document.

could we add a screenshot of a schedule with the 3 parts highlighted to help identify the boundaries of each?

File: app/generator/styles.js [hunk 4]
```diff
@@ -9,7 +9,7 @@ module.exports = {
   header1: {
```
Comment (marctemp) [inline app/generator/styles.js:9]: What's the knock-on effect of these changes for statements?

File: app/messaging/schemas/payment-schedule.js [hunk 5]
```diff
@@ -4,5 +4,6 @@ module.exports = Joi.object({
   order: Joi.number().required(),
   dueDate: Joi.string().optional().allow('', null),
   period: Joi.string().optional().allow('', null),
-  value: Joi.string().required()
+  value: Joi.string().required(),
+  paymentType: Joi.string().required()
```
Comment (marctemp) [inline app/messaging/schemas/payment-schedule.js:8]: can only be 1 of 2 values: "Immediate payment" and "Quarterly payment"

Comment (marctemp) [reply app/messaging/schemas/payment-schedule.js:8]: needs added to the `asyncapi`

File: app/generator/content/schedule/part1/index.js [hunk 6]
```diff
@@ -1,12 +1,21 @@
-const { getAdjustment } = require('./get-adjustment')
-const { getTable } = require('./get-table')
+const path = require('path')
+const { millimetresToPoints } = require('../../../conversion')
+const { getAddress } = require('../../get-address')
+const getAgreementReference = require('./get-agreement-reference')
+const getBusinessName = require('./get-business-name')
+const getSBI = require('./get-sbi')
```
Comment (marctemp) [inline app/generator/content/schedule/part1/index.js:6]: No unit tests for these

File: app/generator/content/schedule/part2/get-reduction-description.js [hunk 7]
```diff
@@ -0,0 +1,20 @@
+const toCurrencyString = require('../../../../generator/to-currency-string')
+
+const getReductionDescription = (schedule) => {
```
Comment (marctemp) [inline app/generator/content/schedule/part2/get-reduction-description.js:3]: No unit tests

File: app/generator/content/schedule/part2/get-top-up-description.js [hunk 8]
```diff
@@ -0,0 +1,20 @@
+const toCurrencyString = require('../../../../generator/to-currency-string')
+
+const getTopUpDescription = (schedule) => {
```
Comment (marctemp) [inline app/generator/content/schedule/part2/get-top-up-description.js:3]: No unit tests

File: app/generator/content/schedule/part3/get-help-info.js [hunk 9]
```diff
@@ -0,0 +1,18 @@
+const getHelpInfo = () => {
```
Comment (marctemp) [inline app/generator/content/schedule/part3/get-help-info.js:1]: No unit tests

File: app/generator/content/schedule/part1/index.js [hunk 10]
```diff
@@ -1,12 +1,21 @@
-const { getAdjustment } = require('./get-adjustment')
-const { getTable } = require('./get-table')
+const path = require('path')
+const { millimetresToPoints } = require('../../../conversion')
+const { getAddress } = require('../../get-address')
+const getAgreementReference = require('./get-agreement-reference')
```
Comment (marctemp) [inline app/generator/content/schedule/part1/index.js:4]: No unit tests

Comment (marctemp) [reply app/generator/content/schedule/part1/index.js:4]: for `getAgreementReference`

File: test/unit/generator/content/schedule/part1/part1.test.js
Comment (marctemp) [inline test/unit/generator/content/schedule/part1/part1.test.js:1]: We need unit tests for part1

File: test/unit/generator/content/schedule/part1/summary.test.js [hunk 11]
```diff
@@ -1,4 +1,4 @@
-const summary = require('../../../../../../app/generator/content/schedule/summary')
+const summary = require('../../../../../../app/generator/content/schedule/part1')
```
[the remark this answers is not in the document]
Comment (marctemp) [reply test/unit/generator/content/schedule/part1/summary.test.js:1]: and it's subsequent describe name

File: test/unit/generator/content/schedule/part1/part1.test.js
Comment (marctemp) [reply test/unit/generator/content/schedule/part1/part1.test.js:1]: think they are moved into summary?

File: test/unit/generator/content/schedule/part2/part2.test.js [hunk 12]
```diff
@@ -0,0 +1,24 @@
+const part2 = require('../../../../../../app/generator/content/schedule/part2')
+const mockSchedule = require('../../../../../mocks/mock-schedule')
+
+describe('schedule part 2', () => {
```
Comment (marctemp) [inline test/unit/generator/content/schedule/part2/part2.test.js:4]: these tests are only for top-ups, need sim. for reductions.

I'd recommend using another set of `describe` to split them up effectively, i.e.

describe('schedule part 2')
describe('when schedule is a top-up')
all the tests you have now
describe('when schedule is a reduction')
sim. tests
describe('when schedule is 0')
new tests, uncovered scenario

File: test/unit/generator/content/schedule/part3/get-table.test.js [hunk 13]
```diff
@@ -73,42 +74,42 @@ describe('get table', () => {
 
   test('header is Payment period for first row', () => {
```
Comment (marctemp) [inline test/unit/generator/content/schedule/part3/get-table.test.js:75]: need to change test names for all these

File: test/unit/generator/content/schedule/part3/get-table.test.js [hunk 14]
```diff
@@ -1,4 +1,5 @@
-const { getTable } = require('../../../../../../app/generator/content/schedule/part1/get-table')
+const { getTable } = require('../../../../../../app/generator/content/schedule/part3/get-table')
+const toCurrencyString = require('../../../../../../app/generator/to-currency-string')
 let mockSchedule
 
 describe('get table', () => {
```
Comment (marctemp) [inline test/unit/generator/content/schedule/part3/get-table.test.js:5]: Can we get these tests again for a reduction schedule?

File: .snyk [hunk 15]
```diff
@@ -6,4 +6,9 @@ ignore:
     - pdfmake:
```
Comment (marctemp) [inline .snyk:6]: missing unit test for the complete output of a schedule when it is a top-up, a reduction and 0 value for `app/generator/content/index.js`

File: test/unit/generator/content/get-address.test.js (hunk 2 above)
Comment (AbidemiAdio) [reply test/unit/generator/content/get-address.test.js:181]: Style measurement unit is millimeters while pdf measurement unit is points. Conversion from millimeters to Points is in multiple of 2.835.

https://github.com/DEFRA/ffc-pay-statement-generator/blob/main/app/generator/conversion.js

File: app/messaging/schemas/payment-schedule.js (hunk 5 above)
Comment (AbidemiAdio) [reply app/messaging/schemas/payment-schedule.js:8]: Asyncapi review ticket already created

File: app/generator/styles.js (hunk 4 above)
Comment (AbidemiAdio) [reply app/generator/styles.js]: Separated common files that required changes, also, common styles that required changes have been separated.

File: app/generator/content/schedule/index.js (hunk 3 above)
Comment (AbidemiAdio) [reply app/generator/content/schedule/index.js:5]: Screenshots now available here 
https://github.com/DEFRA/ffc-pay-statement-generator/tree/163639_Create_top-up_document_template_in_payment-generator_service/acceptance-test/2.1.0/schedule-output

File: app/generator/content/schedule/part1/index.js (hunk 10 above)
Comment (AbidemiAdio) [reply app/generator/content/schedule/part1/index.js:4]: Now added

File: test/unit/generator/content/schedule/part1/summary.test.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (AbidemiAdio) [reply test/unit/generator/content/schedule/part1/summary.test.js]: Corrected

File: test/unit/generator/content/schedule/part3/get-table.test.js (hunk 13 above)
Comment (AbidemiAdio) [reply test/unit/generator/content/schedule/part3/get-table.test.js:75]: changed

File: test/unit/generator/content/schedule/part3/get-table.test.js (hunk 14 above)
Comment (AbidemiAdio) [reply test/unit/generator/content/schedule/part3/get-table.test.js:5]: Added

File: .snyk (hunk 15 above)
Comment (AbidemiAdio) [reply .snyk:6]: as above.

**Review by marctemp — APPROVED** [review]

**Pull request merged**